### PR TITLE
Run KNOWNBUG and THOROUGH regression tests in CI

### PR DESCRIPTION
**Pull request by tautschnig**
Use the check-ubuntu-20_04-cmake-gcc job to run KNOWNBUG (any test
reported as failure will tell us that a bug has unexpectedly been fixed)
an THOROUGH (tests that are expected to pass, but take longer to do so)
tests.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
This is a nice idea.  Is it worth running this once first and PRing any of the KNOWNBUG tests that now work?  My impression is there are probably going to be some and they may need attention to figure out if they are really working on just incorrect tests.

**Comment by NlightNFotis** [comment]
I'm not a huge fan of repurposing existing jobs to be honest.

Would it be possible that this gets done as a separate job?

Otherwise I agree that this is an excellent idea.

**Comment by tautschnig** [comment]
> I'm not a huge fan of repurposing existing jobs to be honest.
> 
> Would it be possible that this gets done as a separate job?
> 
> Otherwise I agree that this is an excellent idea.

I guess I'm still living the times of limited Travis runners :-)  That's fixed now, added two new jobs.

**Comment by tautschnig** [comment]
> This is a nice idea. Is it worth running this once first and PRing any of the KNOWNBUG tests that now work? My impression is there are probably going to be some and they may need attention to figure out if they are really working on just incorrect tests.

Belated response: yes, I'll try to `git bisect` to figure out when exactly tests were fixed. Will keep this PR in "Draft" state for the time being.

**Review by TGWDB — COMMENTED** [review]
I like the idea of the KNOWNBUG checking to routinely check for improvements/changes. The THOROUGH on all PRs seems like it may be a lot of extra load?

File: .github/workflows/pull-request-checks.yaml [hunk 1]
```diff
@@ -170,6 +170,90 @@ jobs:
...
+        with:
+          submodules: recursive
+      - name: Fetch dependencies
+        env:
+          # This is needed in addition to -yq to prevent apt-get from asking for
+          # user input
+          DEBIAN_FRONTEND: noninteractive
+        run: |
+          sudo apt-get update
+          sudo apt-get install --no-install-recommends -yq cmake ninja-build gcc g++ maven flex bison libxml2-utils ccache z3
```
Comment (TGWDB) [inline .github/workflows/pull-request-checks.yaml:186]: The new z3 output on version is not in the new check.

File: .github/workflows/pull-request-checks.yaml [hunk 2]
```diff
@@ -170,6 +170,90 @@ jobs:
...
```
Comment (TGWDB) [inline .github/workflows/pull-request-checks.yaml:233]: z3 version output not here again

**Comment by tautschnig** [comment]
> I like the idea of the KNOWNBUG checking to routinely check for improvements/changes. The THOROUGH on all PRs seems like it may be a lot of extra load?

I'm not sure we care? "CBMC / codecov-coverage-report" takes longer (102 vs 75 minutes), and GitHub happily runs all of the in parallel (up to 20 - according to https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits).

File: .github/workflows/pull-request-checks.yaml (hunk 1 above)
Comment (tautschnig) [reply .github/workflows/pull-request-checks.yaml:186]: Done!

File: .github/workflows/pull-request-checks.yaml (hunk 2 above)
Comment (tautschnig) [reply .github/workflows/pull-request-checks.yaml:233]: Done!

**Review by TGWDB — COMMENTED** [review]
Broadly this PR looks good, but I would prefer that other fixes/changes that are not simply a CI change be a separate PR. For example https://github.com/diffblue/cbmc/pull/5958/commits/00847391c31f6e3fd2781eb379fa244656084420 appears to be bug fix that was found along the way (and great to have done), but not part of the titular change.

**Comment by tautschnig** [comment]
> Broadly this PR looks good, but I would prefer that other fixes/changes that are not simply a CI change be a separate PR. For example [0084739](https://github.com/diffblue/cbmc/commit/00847391c31f6e3fd2781eb379fa244656084420) appears to be bug fix that was found along the way (and great to have done), but not part of the titular change.

Fair point, I've factored those out into #6102, #6103, #6104. The remaining commits could perhaps be factored out as well, but wouldn't be subject to CI-based testing until last commit of this PR is merged (which, in turn, can only be merged if fixing those issues first).

**Comment by tautschnig** [comment]
@TGWDB All dependencies are now merged and this PR is rebased and should be ready for review.

**Review by NlightNFotis — APPROVED** [review]
I'm happy with this personally, provided we don't mark the extra CI jobs as required - they shouldn't be blocking PRs in my opinion.

**Pull request merged**